python/lsst/pipe/tasks/characterizeImage.py

```diff
@@ -0,0 +1,493 @@
+
+        self.display("psf", exposure=dmeRes.exposure, sourceCat=dmeRes.sourceCat)
+
+        # perform final repair with final PSF
+        self.repair.run(exposure=dmeRes.exposure)
+        self.display("repair", exposure=dmeRes.exposure, sourceCat=dmeRes.sourceCat)
+
+        # perform final measurement with final PSF, including measuring and applying aperture correction,
+        # if wanted
+        self.detectAndMeasure.measure(
+            exposure = dmeRes.exposure,
+            exposureIdInfo = exposureIdInfo,
+            sourceCat = dmeRes.sourceCat,
+            allowApCorr = True,  # the default; listed for emphasis
+        )
+        self.display("measure", exposure=dmeRes.exposure, sourceCat=dmeRes.sourceCat)
+
+        return dmeRes
+
+    @pipeBase.timeMethod
+    def detectMeasureAndEstimatePsf(self, exposure, exposureIdInfo, background):
+        """!Perform one iteration of detect, measure and estimate PSF
+
+        Performs the following operations:
+        - if config.doMeasurePsf:
+            - install a simple PSF model (replacing the existing one, if need be)
+        - interpolate over cosmic rays with keepCRs=True
+        - estimate background and subtract it from the exposure
+        - detectAndMeasure: detect, deblend and measure sources, and subtract a refined background model;
+            measuring and applying aperture correction is disabled because we don't have a final PSF
+        - if config.doMeasurePsf:
+            - measure PSF
+
+        @param[in,out] exposure  exposure to characterize (an lsst.afw.image.ExposureF or similar)
+            The following changes are made:
+            - update or set psf
+            - update detection and cosmic ray mask planes
+            - subtract background
+        @param[in] exposureIdInfo  ID info for exposure (an lsst.daf.butlerUtils.ExposureIdInfo)
+        @param[in,out] background  model of background model already subtracted from exposure
+            (an lsst.afw.math.BackgroundList).
+
+        @return pipe_base Struct containing these fields, all from the final iteration
+        of detect sources, measure sources and estimate PSF:
+        - exposure  characterized exposure; image is repaired by interpolating over cosmic rays,
+            mask is updated accordingly, and the PSF model is set
+        - sourceCat  detected sources (an lsst.afw.table.SourceCatalog)
+        - background  model of background subtracted from exposure (an lsst.afw.math.BackgroundList)
+        - psfCellSet  spatial cells of PSF candidates (an lsst.afw.math.SpatialCellSet)
+        """
+        # install a simple PSF model, if wanted
+        if self.config.doMeasurePsf:
+            if self.config.useSimplePsf or not exposure.hasPsf():
+                self.installSimplePsf.run(exposure=exposure)
+        elif not exposure.hasPsf():
+            raise pipeBase.TaskError("exposure has no PSF model and config.doMeasurePsf false")
```

Comment:
Again, don't think `TaskError` is appropriate here because it suppresses the backtrace.